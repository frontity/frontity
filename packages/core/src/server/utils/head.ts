import { HelmetData } from "react-helmet";
import { HeadTags } from "../../../types";

export default (helmet: HelmetData): HeadTags => {
  return {
    base: helmet.base.toString(),
    bodyAttributes: helmet.bodyAttributes.toString(),
    htmlAttributes: helmet.htmlAttributes.toString(),
    link: helmet.link.toString(),
    meta: helmet.meta.toString(),
    noscript: helmet.noscript.toString(),
    script: helmet.script.toString(),
    style: helmet.style.toString(),
    title: helmet.title.toString()
  };
};
