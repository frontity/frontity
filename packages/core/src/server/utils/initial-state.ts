import deepmerge from "deepmerge";
import { NormalizedSettings } from "@frontity/file-settings";
import { Package } from "@frontity/types";

export default ({
  settings,
  url
}: {
  settings: NormalizedSettings;
  url: URL;
}) => {
  let state: Package["state"] = {
    frontity: {
      name: settings.name,
      mode: settings.mode,
      platform: "server",
      initial: {
        path: (/^(.*)page\/\d+\/?$/.exec(url.pathname) || [
          null,
          url.pathname
        ])[1],
        page: parseInt(
          (/^.*page\/(\d+)\/?$/.exec(url.pathname) || [null, "1"])[1],
          10
        )
      },
      packages: settings.packages.map(pkg => pkg.name)
    }
  };

  // Merge the initial state with the general state of settings.
  state = deepmerge(settings.state, state, {
    clone: false
  });

  // Merge the state with each package state, in order.
  settings.packages.forEach(pkg => {
    state = deepmerge(state, pkg.state, {
      clone: false
    });
  });

  return state;
};
