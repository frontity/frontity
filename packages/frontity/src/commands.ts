#! /usr/bin/env node
import * as tsNode from "ts-node";

tsNode.register({
  transpileOnly: true,
  compilerOptions: {
    // Target latest version of ECMAScript.
    target: "es2017",
    // Search under node_modules for non-relative imports.
    moduleResolution: "node",
    // commonjs modules.
    module: "commonjs",
    // Allow default imports from modules with no default export.
    allowSyntheticDefaultImports: true,
    // Don't emit; allow Babel to transform files.
    noEmit: true,
    // Import non-ES modules as default imports.
    esModuleInterop: true,
    // Resolve JSON files.
    resolveJsonModule: true,
    // Support for JSX.
    jsx: "react",
    // Transpile JS as well.
    allowJs: true
  }
});

import program from "commander";
import { readFileSync } from "fs-extra";
import { resolve } from "path";
import {
  create,
  createPackage,
  dev,
  build,
  serve,
  subscribe,
  info
} from "./actions";

const { version } = JSON.parse(
  readFileSync(resolve(__dirname, "../package.json"), { encoding: "utf8" })
);

// Sets the version and the description of the program.
program.version(version).description("Frontity CLI");

// Registers a `create` command that takes an optional
// parameter called `name`. It also accepts the following
// options: --typescript, --use-cwd.
program
  .command("create [name]")
  .option("-t, --typescript", "Adds support for TypeScript")
  .option("-c, --use-cwd", "Generates the project in the current directory.")
  .description("Creates a new Frontity project.")
  .action(create);

program
  .command("create-package [name]")
  .description("Creates a new Frontity package.")
  .action(createPackage);

program
  .command("dev")
  .option("-p, --production", "Builds the project for production.")
  .option("--port <port>", "Runs the server on a custom port. Default is 3000.")
  .option("-s, --https", "Runs the server using https.")
  .option("--target <target>")
  .description("Starts a server in development mode.")
  .action(dev);

program
  .command("build")
  .option("-d, --development", "Builds the project for development.")
  .option("--target <target>")
  .description("Builds the project for production.")
  .action(build);

program
  .command("serve")
  .option("--port <port>", "Runs the server on a custom port. Default is 3000.")
  .option("-s, --https", "Runs the server using https.")
  .description("Starts a server in production mode.")
  .action(serve);

program
  .command("subscribe <email>")
  .description("Subscribe to Frontity newsletter.")
  .action(subscribe);

program
  .command("info")
  .description("Get environment information for debugging and issue reporting.")
  .action(info);

// Parses the arguments and adds them to the `command` object.
program.parse(process.argv);
