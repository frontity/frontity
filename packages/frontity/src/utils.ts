import chalk from "chalk";
import { readdir as readDir } from "fs-extra";

export const isPackageNameValid = (name: string): boolean => {
  if (name.length >= 214) return false;
  if ("._".includes(name[0])) return false;
  const specialCharacters = /[/:*?"<>|]/;
  if (specialCharacters.test(name)) return false;
  const nameConventionMatch = /^(?:@[a-z0-9_.-]+\/)?[a-z0-9_.-]+$/;
  return nameConventionMatch.test(name);
};

export const isThemeNameValid = (name: string): boolean => {
  const nameConventionMatch = /^(?:@[\w-]+\/)?[\w-]+$/;
  return nameConventionMatch.test(name);
};

export const errorLogger = (error: Error, message?: string) => {
  console.error(
    chalk.bold.red("\nError: ") +
      chalk.red(error.message) +
      "\n\n" +
      (message || "") +
      `If you need help please visit ${chalk.underline.magenta(
        "https://community.frontity.org/"
      )}.\n`
  );
  process.exit(1);
};

export const isFrontityProjectRoot = async (path: string) => {
  const dirContent = await readDir(path);
  return dirContent.some(content =>
    /^frontity\.settings\.(js|ts)$/i.test(content)
  );
};
