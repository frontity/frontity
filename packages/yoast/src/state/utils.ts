import { YoastSocialDefaults, YoastMeta } from "../../types";
import { Author, Taxonomy, PostType, Attachment } from "@frontity/source/";

export type Entity = (Author | Taxonomy | PostType | Attachment) & {
  yoast_meta: YoastMeta;
};

export const getEntity = ({ state, data }): Entity => {
  if (data.isPostType) return state.source[data.type][data.id];
  if (data.isTaxonomy) return state.source[data.taxonomy][data.id];
  if (data.isAuthor) return state.source.author[data.id];
  return null;
};

export const appendPage = ({
  title,
  page,
  totalPages
}: {
  title: string;
  page: number;
  totalPages: number;
}) => (page > 1 ? `${title} - Page ${page} of ${totalPages}` : title);

export const getSocialDefaults = ({ state }): YoastSocialDefaults => {
  const data = state.source.get(state.router.link);

  let type: string;
  let id: number;

  if (data.isArchive) ({ type, id } = data.items[0]);
  if (data.isPostType) ({ type, id } = data);

  if (type && id)
    return state.source[type][id].yoast_meta.yoast_wpseo_social_defaults;

  return {};
};

export const getPathname = route => {
  const [, pathname] = /^(?:(?:[^:/?#]+):)?(?:\/\/(?:[^/?#]*))?([^?#]*)/.exec(
    route
  );
  return pathname;
};

export const getPage = route => {
  const [, page] = /^(?:.*)page\/(\d+)\/?(\?.*)?$/.exec(getPathname(route)) || [
    null,
    "1"
  ];
  return parseInt(page, 10);
};

export const hasName = (
  entity: Entity
): entity is (Author | Taxonomy) & {
  yoast_meta: YoastMeta;
} => "name" in entity;
