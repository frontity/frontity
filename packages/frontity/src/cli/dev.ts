import devCommand from "../commands/dev";
import { prompt } from "inquirer";
import { errorLogger } from "../utils";

/**
 * The options for the {@link build} command.
 */
interface DevOptions {
  /**
   * Describes the environments the final bundles will support in the client
   * side, which can be either `es5` for old browsers with ES6 Proxy support or
   * `module` for modern browsers with `<script type="module"></script>`
   * support.
   *
   * - Module: edge \>= 16, firefox \>= 60, chrome \>= 61, safari \>= 10.1,
   * opera \>= 48, ios_saf \>= 10.3, and_ff \>= 60.
   *
   * - ES5: and_chr \>= 67, and_ff \>= 18, and_uc \>= 11.8, android \>= 67, not
   * android \<= 4.4.4, chrome \>= 49, edge \>= 12, firefox \>= 18, ios_saf \>=
   * 10, not op_mini all, op_mob \>= 46, opera \>= 36, "safari \>= 10, samsung
   * \>= 5.
   *
   * It can be also configured using the `FRONTITY_DEV_TARGET` env variable.
   *
   * @defaultValue "module"
   */
  target?: string;

  /**
   * Indicates if the bundles should be generated using the mode="production"
   * option of Webpack.
   *
   * Webpack docs: https://webpack.js.org/configuration/mode/.
   *
   * It can be also configured using the `FRONTITY_DEV_PRODUCTION` env
   * variable.
   *
   * @defaultValue false
   */
  production?: boolean;

  /**
   * The publicPath option of Webpack.
   *
   * Webpack docs: https://webpack.js.org/guides/public-path/.
   *
   * It can be also configured using the `FRONTITY_DEV_PUBLIC_PATH` env
   * variable.
   *
   * @example
   * ```sh
   * npx frontity dev --public-path /custom/path
   * npx frontity dev --public-path http://cdn.domain.com/custom/path
   * ```
   *
   * @defaultValue "/static/"
   */
  publicPath?: string;

  /**
   * The port that should be used to open the server.
   *
   * It can be also configured using the `FRONTITY_DEV_PORT` env variable.
   *
   * @example
   * ```sh
   * # This will start Frontity at http://localhost:3003.
   * npx frontity dev --port 3003
   * ```
   *
   * @defaultValue "3000"
   */
  port?: string;

  /**
   * Indicates if the server should be started in HTTPS mode.
   *
   * The certificates used are stored internally in the Frontity core and are
   * not meant to be used in production, only in local development.
   *
   * It can be also configured using the `FRONTITY_DEV_HTTPS` env variable.
   *
   * @example
   * ```sh
   * # This will start Frontity in https://localhost:3000.
   * npx frontity dev --https
   * ```
   *
   * @defaultValue false
   */
  https?: boolean;

  /**
   * Indicates if a browser tab should be opened once Frontity starts.
   *
   * It can be also configured using the `FRONTITY_DEV_DONT_OPEN_BROWSER` env
   * variable.
   *
   * @defaultValue false
   */
  dontOpenBrowser?: boolean;

  /**
   * If active, it creates HTML files for bundle analyzing inside the
   * /build/analyze/ folder.
   *
   * It can be also configured using the `FRONTITY_DEV_ANALYZE` env variable.
   *
   * @defaultValue false
   */
  analyze?: boolean;

  /**
   * The name of the configuration to run.
   */
  siteName?: string;
}

/**
 * The dev CLI command, usually run with `npx frontity dev`.
 *
 * It takes args from the CLI and checks for the presence of environment
 * variables. Then, it runs the dev command programatically.
 *
 * @param options - Defined in {@link DevOptions}.
 */
const dev = async ({
  target = process.env.FRONTITY_DEV_TARGET || "module",
  port = process.env.FRONTITY_DEV_PORT || "3000",
  https = !!process.env.FRONTITY_DEV_HTTPS,
  production = !!process.env.FRONTITY_DEV_PRODUCTION,
  publicPath = process.env.FRONTITY_DEV_PUBLIC_PATH || "/static/",
  dontOpenBrowser = !!process.env.FRONTITY_DEV_DONT_OPEN_BROWSER,
  analyze = !!process.env.FRONTITY_DEV_ANALYZE,
  siteName = process.env.FRONTITY_DEV_SITE_NAME,
}: DevOptions) => {
  // Check `target` parameter.
  if (target && target !== "es5" && target !== "module") {
    errorLogger(
      new Error(
        `The target specified is invalid: "${target}". Use either "module" or "es5".`
      )
    );
  }

  // Check `port` parameter.
  if (port && Number.isNaN(parseInt(port, 10))) {
    errorLogger(new Error(`The port number specified is not valid: ${port}.`));
  }

  /**
   * The make function for command.
   *
   * @param args - The custom valued args.
   * @returns Nothing.
   */
  const make = async ({ name }) =>
    await devCommand({
      target: target as "es5" | "module",
      port: parseInt(port, 10),
      production,
      https,
      publicPath,
      dontOpenBrowser,
      analyze,
      siteName: name,
    });

  /**
   * The main program.
   *
   * @param args - The custom value args to pass.
   */
  const main = async ({ name }) => {
    try {
      await make({ name });
    } catch (e) {
      const sites = e.sites.map((site) => site.name);
      const options = [
        {
          type: "list",
          name: "siteName",
          message:
            "No '--site-name' parameter passed to the dev command. Please select one:",
          choices: sites,
          default: sites[0],
        },
      ];
      const results = await prompt(options);

      main({ name: results.siteName });
    }
  };

  await main({ name: siteName });
};

export default dev;
