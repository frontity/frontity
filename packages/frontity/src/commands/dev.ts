import chalk from "chalk";
import { errorLogger } from "../utils";
import choosePort from "../utils/choosePort";

const HOST = process.env.HOST || "0.0.0.0";

/**
 * Options for the {@link devCommand} function.
 */
interface DevOptions {
  /**
   * Builds the project for production.
   *
   * @defaultValue `false`
   */
  production?: boolean;

  /**
   * Runs the server on a custom port.
   *
   * @defaultValue `3000`
   */
  port?: number;

  /**
   * Runs the server using https.
   *
   * @defaultValue `false`
   */
  https?: boolean;

  /**
   * Create bundles with "es5" or "module".
   *
   * @defaultValue `"module"`
   */
  target?: "es5" | "module";

  /**
   * Don't open a browser window with the localhost.
   *
   * @defaultValue `false`
   */
  dontOpenBrowser?: boolean;

  /**
   * Set the public path for static assets.
   *
   * @defaultValue `"/static/"`
   */
  publicPath?: string;

  /**
   * Create HTML files for bundle analyzing.
   *
   * @defaultValue `false`
   */
  analyze?: boolean;

  /**
   * The name of the site you want to run.
   */
  siteName?: string;
}

/**
 * Start a server in development mode.
 *
 * This function is executed by the CLI when running the `npx frontity dev`
 * command.
 *
 * @param options - Object of type {@link DevOptions}.
 */
const devCommand = async ({
  production = false,
  port = 3000,
  https = false,
  target = "module",
  dontOpenBrowser = false,
  publicPath = "/static/",
  analyze = false,
  siteName,
}: DevOptions) => {
  // Try getting the `dev` function from `@frontity/core`.
  let dev: (...options: any[]) => Promise<void>;

  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    dev = require("@frontity/core").dev;
  } catch (error) {
    const message =
      `Make sure that you are running ${chalk.green(
        "frontity"
      )} inside a Frontity project.\n` +
      `If so try installing ${chalk.green(
        "@frontity/core"
      )} again with ${chalk.green("npm i @frontity/core")}.\n`;
    errorLogger(error, message);
  }

  // Generate options for the core's `dev` function.
  const options = {
    mode: production ? "production" : "development",
    port,
    isHttps: !!https,
    target,
    openBrowser: !dontOpenBrowser,
    publicPath,
    analyze,
    siteName,
  };

  try {
    const port = await choosePort(HOST, options.port);
    if (port === null) {
      return;
    }
    await dev({ ...options, port });
  } catch (error) {
    if (error.sites) {
      throw error;
    }

    errorLogger(error);
  }
};

export default devCommand;
