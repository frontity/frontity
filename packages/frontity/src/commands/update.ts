import { EventPromised } from "../utils/eventPromised";
import { updateProject } from "../steps/update";

/**
 * The type of events that the EventPromised instance accepts.
 */
type EventTypes = "error" | "message";

/**
 * The update command, exported to be used programatically.
 *
 * @returns An instance of {@link EventPromised}, which is a promise that can
 * also send events using an EventEmitter.
 */
const update = () =>
  new EventPromised<EventTypes>((resolve, reject, emit) => {
    (async () => {
      {
        try {
          // Small hack so the first `emit` message is skipped.
          await new Promise((resolve) => setTimeout(resolve));

          emit("message", "\nUpdating Frontity root project\n");
          await updateProject();

          // Update the root project.
          // let step = updateProject();
          // emit("message", "Updating the root project", step);
          // await step;
          // const packages = await step;
          // Object.entries(packages).forEach(([pkg, version]) => {
          //   emit("message", `Updated ${pkg} to ${version}.`);
          // });

          // Update the local packages.
          // step = updatePackages();
          // emit("message", "Updating the root project", step);
          // await step;
        } catch (error) {
          reject(error);
        }
      }
    })().then(resolve);
  });

export default update;
