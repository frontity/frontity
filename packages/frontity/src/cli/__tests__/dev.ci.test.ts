import dev from "../dev";
import * as utils from "../../utils";
import * as devCmd from "../../commands/dev";
import { EventPromised } from "../../utils/eventPromised";

jest.mock("../../commands/dev");

const mockedUtils = utils as any;
const mockedDev = devCmd as jest.Mocked<typeof devCmd>;

describe("dev", () => {
  beforeEach(() => {
    mockedUtils.isFrontityProjectRoot = jest.fn();
    mockedUtils.errorLogger = jest.fn(() => {
      throw new Error("Process terminated");
    });
    mockedDev.default.mockReset();
    mockedDev.default.mockReturnValue(
      new EventPromised((resolve) => resolve())
    );
  });

  test("should receive default values", async () => {
    await dev({});
    expect(mockedDev.default.mock.calls).toMatchSnapshot();
  });

  test("should get values from ENV variables", async () => {
    process.env.FRONTITY_DEV_TARGET = "es5";
    process.env.FRONTITY_DEV_PORT = "3001";
    process.env.FRONTITY_DEV_HTTPS = "true";
    process.env.FRONTITY_DEV_PRODUCTION = "true";
    process.env.FRONTITY_DEV_PUBLIC_PATH = "/public/path";
    process.env.FRONTITY_DEV_DONT_OPEN_BROWSER = "true";
    process.env.FRONTITY_DEV_ANALYZE = "true";
    process.env.FRONTITY_DEV_SITE_NAME = "siteName";

    await dev({});
    expect(mockedDev.default.mock.calls).toMatchSnapshot();
  });

  test("should ignore ENV variables for passed arguments", async () => {
    process.env.FRONTITY_DEV_TARGET = "es5";
    process.env.FRONTITY_DEV_PORT = "3001";
    process.env.FRONTITY_DEV_HTTPS = "true";
    process.env.FRONTITY_DEV_PRODUCTION = "true";
    process.env.FRONTITY_DEV_PUBLIC_PATH = "/public/path";
    process.env.FRONTITY_DEV_DONT_OPEN_BROWSER = "true";
    process.env.FRONTITY_DEV_ANALYZE = "true";
    process.env.FRONTITY_DEV_SITE_NAME = "siteName";

    await dev({
      target: "module",
      port: "4000",
      https: false,
      production: false,
      publicPath: "/static",
      dontOpenBrowser: false,
      analyze: false,
      siteName: "customSiteName",
    });
    expect(mockedDev.default.mock.calls).toMatchSnapshot();
  });

  test("should fail if an invalid target is passed", async () => {
    try {
      await dev({ target: "not-valid-target" });
    } catch (error) {
      // Ignore error.
    }
    expect(mockedDev.default).not.toHaveBeenCalled();
    expect(mockedUtils.errorLogger.mock.calls).toMatchSnapshot();
  });

  test("should fail if an invalid port is passed", async () => {
    try {
      await dev({ port: "not-valid-port" });
    } catch (error) {
      // Ignore error.
    }
    expect(mockedDev.default).not.toHaveBeenCalled();
    expect(mockedUtils.errorLogger.mock.calls).toMatchSnapshot();
  });
});
