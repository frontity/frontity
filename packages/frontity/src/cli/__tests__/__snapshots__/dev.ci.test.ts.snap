// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dev should fail if an invalid port is passed 1`] = `
Array [
  Array [
    [Error: The port number specified is not valid: not-valid-port.],
  ],
]
`;

exports[`dev should fail if an invalid target is passed 1`] = `
Array [
  Array [
    [Error: The target specified is invalid: "not-valid-target". Use either "module" or "es5".],
  ],
]
`;

exports[`dev should get values from ENV variables 1`] = `
Array [
  Array [
    Object {
      "analyze": true,
      "dontOpenBrowser": true,
      "https": true,
      "port": 3001,
      "production": true,
      "publicPath": "/public/path",
      "siteName": "siteName",
      "target": "es5",
    },
  ],
]
`;

exports[`dev should ignore ENV variables for passed arguments 1`] = `
Array [
  Array [
    Object {
      "analyze": false,
      "dontOpenBrowser": false,
      "https": false,
      "port": 4000,
      "production": false,
      "publicPath": "/static",
      "siteName": "customSiteName",
      "target": "module",
    },
  ],
]
`;

exports[`dev should receive default values 1`] = `
Array [
  Array [
    Object {
      "analyze": false,
      "dontOpenBrowser": false,
      "https": false,
      "port": 3000,
      "production": false,
      "publicPath": "/static/",
      "siteName": undefined,
      "target": "module",
    },
  ],
]
`;
